src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -329,22 +335,52 @@ await CallWithRequestTracing(
                                 }
                             }
 
-                            if (refreshAll)
+                            if (!refreshAll)
+                            {
+                                watchedSelectedKeyValueCollections = _watchedSelectedKeyValueCollections;
+
+                                watchedFeatureFlagCollections = _watchedFeatureFlagCollections;
+
+                                if (_options.RefreshOptions.RegisterAllEnabled)
+                                {
+                                    bool selectedKeyValueCollectionsChanged = await UpdateWatchedCollections(refreshableSelectedKeyValueWatchers, watchedSelectedKeyValueCollections, client, cancellationToken).ConfigureAwait(false);
+
+                                    if (!selectedKeyValueCollectionsChanged)
+                                    {
+                                        logDebugBuilder.AppendLine(LogHelper.BuildSelectedKeyValueCollectionsUnchangedMessage(endpoint.ToString()));
+                                    }
+                                    else
+                                    {
+                                        keyValueChanges.AddRange(await GetRefreshedKeyValueCollections(_selectedKeyValueWatchers, watchedSelectedKeyValueCollections, _mappedData, client, cancellationToken).ConfigureAwait(false));
+
+                                        logInfoBuilder.Append(LogHelper.BuildSelectedKeyValueCollectionsUpdatedMessage());
+                                    }
+                                }
+
+                                bool featureFlagCollectionsChanged = await UpdateWatchedCollections(refreshableFeatureFlagWatchers, watchedFeatureFlagCollections, client, cancellationToken).ConfigureAwait(false);
+
+                                if (!featureFlagCollectionsChanged)
+                                {
+                                    logDebugBuilder.AppendLine(LogHelper.BuildFeatureFlagsUnchangedMessage(endpoint.ToString()));
+                                }
+                                else
+                                {
+                                    keyValueChanges.AddRange(await GetRefreshedKeyValueCollections(_options.FeatureFlagWatchers, watchedFeatureFlagCollections, _mappedData, client, cancellationToken).ConfigureAwait(false));
+
+                                    logInfoBuilder.Append(LogHelper.BuildFeatureFlagsUpdatedMessage());
+                                }
+                            }
+                            else
                             {
                                 // Trigger a single load-all operation if a change was detected in one or more key-values with refreshAll: true
-                                data = await LoadSelectedKeyValues(client, cancellationToken).ConfigureAwait(false);
+                                // or if RegisterAll was called and any loaded key-value changed
```

Comment:
Do we reload key-values when feature flags change? And vice versa?

Reply:
I think we should treat key value and feature flag respectively. `RefreshOption.Register` or `RefreshOption.RegisterAll` only affect key value refresh. Feature flag has its own refresh behavior, which is "RegisterAll" by default. This is how JavaScript provider does.

Reply:
Oh I left this comment there on accident from a previous iteration, yes what @zhiyuanliang-ms said is right. That else statement is only for `refreshAll` which is from sentinel key.